src/metabase/api/card.clj

```diff
@@ -104,6 +106,24 @@
        ;; now check if model-id really occurs as a card ID
        (filter (fn [card] (some #{model-id} (-> card :dataset_query query/collect-card-ids))))))
 
+(defn- cards-for-segment-or-metric
+  [model-type model-id]
+  (->> (t2/select :model/Card {:where [:like :dataset_query (str "%" (name model-type) "%" model-id "%")]})
+       ;; now check if the segment/metric with model-id really occurs in a filter/aggregation expression
+       (filter (fn [card]
+                 (let [query (-> card :dataset_query lib.convert/->pMBQL)]
```

Comment:
This is relying on `dataset_query` having been normalized when written into the DB.

Reply:
A single conversion shouldn't be bad, especially when compared with the overhead of DB and network calls.